ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -96,18 +103,36 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
    * @param size the number of bytes to upload.
    * @throws CloudStorageException if the upload failed.
    */
-  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size) throws CloudStorageException {
-    boolean performUpload = messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !messageInfo.isDeleted();
-    // May need to encrypt buffer before upload
-    if (performUpload) {
+  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
+      throws CloudStorageException, IOException {
+    if (shouldUpload(messageInfo)) {
       BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      if (cloudConfig.vcrRequireEncryption && !BlobId.isEncrypted(blobId.getID())) {
```

Comment:
fyi one change that I'll have to make is that because the blobID encrypted field is unreliable, I will be parsing the messageBuf object to get the isEncryption information from the message header and use that to make the decision instead

Reply:
yes. On frontend static method isEncrypted() is reliable. On server/vcr,  isEncrypted() not valid for v3.